// ng-circle-progress/esm2015/lib/ng-circle-progress.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CircleProgressComponent, CircleProgressOptions } from './ng-circle-progress.component';
import * as ɵngcc0 from '@angular/core';
export class NgCircleProgressModule {
    static forRoot(options = {}) {
        return {
            ngModule: NgCircleProgressModule,
            providers: [
                { provide: CircleProgressOptions, useValue: options }
            ]
        };
    }
}
NgCircleProgressModule.ɵfac = function NgCircleProgressModule_Factory(t) { return new (t || NgCircleProgressModule)(); };
NgCircleProgressModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgCircleProgressModule });
NgCircleProgressModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgCircleProgressModule, [{
        type: NgModule,
        args: [{
                declarations: [CircleProgressComponent],
                imports: [
                    CommonModule
                ],
                exports: [CircleProgressComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgCircleProgressModule, { declarations: function () { return [CircleProgressComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [CircleProgressComponent]; } }); })();
