// ng-bootstrap-form-validation/esm2015/lib/Directives/form-control.directive.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Input, HostBinding, Optional, Host, SkipSelf, Inject } from "@angular/core";
import { ControlContainer, FormControl } from "@angular/forms";
import { BootstrapVersion } from "../Enums/BootstrapVersion";
import { BOOTSTRAP_VERSION } from "../Tokens/tokens";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
export function controlPath(name, parent) {
    // tslint:disable-next-line:no-non-null-assertion
    return [...parent.path, name];
}
let FormControlDirective = class FormControlDirective {
    constructor(
    // this value might be null, but we union type it as such until
    // this issue is resolved: https://github.com/angular/angular/issues/25544
    parent, bootstrapVersion) {
        this.parent = parent;
        this.bootstrapVersion = bootstrapVersion;
    }
    get validClass() {
        if (!this.control) {
            return false;
        }
        return (this.bootstrapFour &&
            this.control.valid &&
            (this.control.touched || this.control.dirty));
    }
    get invalidClass() {
        if (!this.control) {
            return false;
        }
        return (this.bootstrapFour &&
            this.control.invalid &&
            this.control.touched &&
            this.control.dirty);
    }
    get path() {
        return controlPath(this.formControlName, this.parent);
    }
    get control() {
        return this.formDirective && this.formDirective.getControl(this);
    }
    get formDirective() {
        return this.parent ? this.parent.formDirective : null;
    }
    get bootstrapFour() {
        return this.bootstrapVersion === BootstrapVersion.Four;
    }
};
FormControlDirective.ɵfac = function FormControlDirective_Factory(t) { return new (t || FormControlDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ControlContainer, 13), ɵngcc0.ɵɵdirectiveInject(BOOTSTRAP_VERSION)); };
FormControlDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: FormControlDirective, selectors: [["", 8, "form-control"], ["", 8, "form-check-input"], ["", 8, "custom-control-input"]], hostVars: 4, hostBindings: function FormControlDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("is-valid", ctx.validClass)("is-invalid", ctx.invalidClass);
    } }, inputs: { formControlName: "formControlName", formControl: "formControl" } });
FormControlDirective.ctorParameters = () => [
    { type: ControlContainer, decorators: [{ type: Optional }, { type: Host }, { type: SkipSelf }] },
    { type: undefined, decorators: [{ type: Inject, args: [BOOTSTRAP_VERSION,] }] }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], FormControlDirective.prototype, "formControlName", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], FormControlDirective.prototype, "formControl", void 0);
__decorate([
    HostBinding("class.is-valid"),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], FormControlDirective.prototype, "validClass", null);
__decorate([
    HostBinding("class.is-invalid"),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], FormControlDirective.prototype, "invalidClass", null);
FormControlDirective = __decorate([ __param(0, Optional()),
    __param(0, Host()),
    __param(0, SkipSelf()),
    __param(1, Inject(BOOTSTRAP_VERSION)),
    __metadata("design:paramtypes", [ControlContainer, Number])
], FormControlDirective);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormControlDirective, [{
        type: Directive,
        args: [{
                // tslint:disable-next-line:directive-selector
                selector: ".form-control,.form-check-input,.custom-control-input"
            }]
    }], function () { return [{ type: ɵngcc1.ControlContainer, decorators: [{
                type: Optional
            }, {
                type: Host
            }, {
                type: SkipSelf
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [BOOTSTRAP_VERSION]
            }] }]; }, { validClass: [{
            type: HostBinding,
            args: ["class.is-valid"]
        }], invalidClass: [{
            type: HostBinding,
            args: ["class.is-invalid"]
        }], formControlName: [{
            type: Input
        }], formControl: [{
            type: Input
        }] }); })();
export { FormControlDirective };
