// @angular-redux/store/lib/testing/ng-redux-testing.module.js
"use strict";
var ɵngcc0 = require('@angular/core');
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var store_1 = require("@angular-redux/store");
var ng_redux_mock_1 = require("./ng-redux.mock");
var dev_tools_mock_1 = require("./dev-tools.mock");
// Needs to be initialized early so @select's use the mocked version too.
var mockNgRedux = ng_redux_mock_1.MockNgRedux.getInstance();
/** @hidden */
function _mockNgReduxFactory() {
    return mockNgRedux;
}
exports._mockNgReduxFactory = _mockNgReduxFactory;
var NgReduxTestingModule = /** @class */ (function () {
    function NgReduxTestingModule() {
    }
NgReduxTestingModule.ɵfac = function NgReduxTestingModule_Factory(t) { return new (t || NgReduxTestingModule)(); };
NgReduxTestingModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgReduxTestingModule });
NgReduxTestingModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        { provide: store_1.NgRedux, useFactory: _mockNgReduxFactory },
        { provide: store_1.DevToolsExtension, useClass: dev_tools_mock_1.MockDevToolsExtension },
    ], imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgReduxTestingModule, [{
        type: core_1.NgModule,
        args: [{
                imports: [],
                providers: [
                    { provide: store_1.NgRedux, useFactory: _mockNgReduxFactory },
                    { provide: store_1.DevToolsExtension, useClass: dev_tools_mock_1.MockDevToolsExtension },
                ]
            }]
    }], function () { return []; }, null); })();
    return NgReduxTestingModule;
}());
exports.NgReduxTestingModule = NgReduxTestingModule;
