// ng-bootstrap-form-validation/esm5/lib/Directives/form-validation.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, EventEmitter, Input, Output, HostListener } from "@angular/core";
import { FormArray, FormControl, FormGroup } from "@angular/forms";
import * as ɵngcc0 from '@angular/core';
var FormValidationDirective = /** @class */ (function () {
    function FormValidationDirective() {
        this.validSubmit = new EventEmitter();
    }
    FormValidationDirective.prototype.onSubmit = function () {
        this.markAsTouchedAndDirty(this.formGroup);
        if (this.formGroup.valid) {
            this.validSubmit.emit(this.formGroup.value);
        }
    };
    FormValidationDirective.prototype.markAsTouchedAndDirty = function (control) {
        var _this = this;
        if (control instanceof FormGroup) {
            Object.keys(control.controls).forEach(function (key) {
                return _this.markAsTouchedAndDirty(control.controls[key]);
            });
        }
        else if (control instanceof FormArray) {
            control.controls.forEach(function (c) { return _this.markAsTouchedAndDirty(c); });
        }
        else if (control instanceof FormControl && control.enabled) {
            control.markAsDirty();
            control.markAsTouched();
            control.updateValueAndValidity();
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", FormGroup)
    ], FormValidationDirective.prototype, "formGroup", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], FormValidationDirective.prototype, "validSubmit", void 0);
    __decorate([
        HostListener("submit"),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], FormValidationDirective.prototype, "onSubmit", null);
FormValidationDirective.ɵfac = function FormValidationDirective_Factory(t) { return new (t || FormValidationDirective)(); };
FormValidationDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: FormValidationDirective, selectors: [["", "formGroup", ""]], hostBindings: function FormValidationDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("submit", function FormValidationDirective_submit_HostBindingHandler() { return ctx.onSubmit(); });
    } }, inputs: { formGroup: "formGroup" }, outputs: { validSubmit: "validSubmit" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormValidationDirective, [{
        type: Directive,
        args: [{
                // tslint:disable-next-line:directive-selector
                selector: "[formGroup]"
            }]
    }], function () { return []; }, { validSubmit: [{
            type: Output
        }], onSubmit: [{
            type: HostListener,
            args: ["submit"]
        }], formGroup: [{
            type: Input
        }] }); })();
    return FormValidationDirective;
}());
export { FormValidationDirective };
