// ngx-countup/esm2015/lib/count-up.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CountUpDirective } from './count-up.directive';
import * as ɵngcc0 from '@angular/core';
let CountUpModule = class CountUpModule {
};
CountUpModule.ɵfac = function CountUpModule_Factory(t) { return new (t || CountUpModule)(); };
CountUpModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CountUpModule });
CountUpModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CountUpModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [CountUpDirective],
                exports: [CountUpDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CountUpModule, { declarations: function () { return [CountUpDirective]; }, exports: function () { return [CountUpDirective]; } }); })();
export { CountUpModule };
