// ngx-trend/esm2015/trend/trend.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TrendComponent } from './trend.component';
import * as ɵngcc0 from '@angular/core';
export class TrendModule {
}
TrendModule.ɵfac = function TrendModule_Factory(t) { return new (t || TrendModule)(); };
TrendModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TrendModule });
TrendModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TrendModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [TrendComponent],
                declarations: [TrendComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TrendModule, { declarations: function () { return [TrendComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [TrendComponent]; } }); })();
