// ngx-textarea-autosize/esm2015/autosize.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostListener, Input, HostBinding, Renderer2 } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let AutosizeDirective = class AutosizeDirective {
    constructor(elem, renderer) {
        this.elem = elem;
        this.renderer = renderer;
        this.overflow = 'hidden';
        this.rows = 1;
    }
    ngAfterViewInit() {
        this.resize();
    }
    ngDoCheck() {
        this.resize();
    }
    resize() {
        const textarea = this.elem.nativeElement;
        // Calculate border height which is not included in scrollHeight
        const borderHeight = textarea.offsetHeight - textarea.clientHeight;
        // Reset textarea height to auto that correctly calculate the new height
        this.setHeight('auto');
        // Set new height
        this.setHeight(`${textarea.scrollHeight + borderHeight}px`);
    }
    setHeight(value) {
        this.renderer.setStyle(this.elem.nativeElement, 'height', value);
    }
};
AutosizeDirective.ɵfac = function AutosizeDirective_Factory(t) { return new (t || AutosizeDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
AutosizeDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: AutosizeDirective, selectors: [["textarea", "autosize", ""]], hostVars: 3, hostBindings: function AutosizeDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("input", function AutosizeDirective_input_HostBindingHandler() { return ctx.resize(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("rows", ctx.rows);
        ɵngcc0.ɵɵstyleProp("overflow", ctx.overflow);
    } }, inputs: { rows: "rows" } });
tslib_1.__decorate([
    HostBinding('style.overflow'),
    tslib_1.__metadata("design:type", Object)
], AutosizeDirective.prototype, "overflow", void 0);
tslib_1.__decorate([
    Input(),
    HostBinding('rows'),
    tslib_1.__metadata("design:type", Object)
], AutosizeDirective.prototype, "rows", void 0);
tslib_1.__decorate([
    HostListener('input'),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", void 0)
], AutosizeDirective.prototype, "resize", null);
AutosizeDirective = tslib_1.__decorate([ tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], AutosizeDirective);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AutosizeDirective, [{
        type: Directive,
        args: [{
                selector: 'textarea[autosize]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { overflow: [{
            type: HostBinding,
            args: ['style.overflow']
        }], rows: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['rows']
        }], resize: [{
            type: HostListener,
            args: ['input']
        }] }); })();
export { AutosizeDirective };
