// ng-bootstrap-form-validation/esm2015/lib/Services/error-message.service.js
import { __decorate, __metadata, __param } from "tslib";
import { Injectable, Inject } from "@angular/core";
import { DEFAULT_ERRORS } from "../default-errors";
import { CUSTOM_ERROR_MESSAGES } from "../Tokens/tokens";
import * as i0 from "@angular/core";
import * as i1 from "../Tokens/tokens";
import * as ɵngcc0 from '@angular/core';
let ErrorMessageService = class ErrorMessageService {
    constructor(customErrorMessages) {
        this.customErrorMessages = customErrorMessages;
        this.defaultErrors = DEFAULT_ERRORS;
        this.errorMessages = customErrorMessages.reduce((acc, cur) => acc.concat(cur), this.defaultErrors);
    }
};
ErrorMessageService.ɵfac = function ErrorMessageService_Factory(t) { return new (t || ErrorMessageService)(ɵngcc0.ɵɵinject(CUSTOM_ERROR_MESSAGES)); };
ErrorMessageService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Inject, args: [CUSTOM_ERROR_MESSAGES,] }] }
];
ErrorMessageService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ErrorMessageService_Factory() { return new ErrorMessageService(i0.ɵɵinject(i1.CUSTOM_ERROR_MESSAGES)); }, token: ErrorMessageService, providedIn: "root" });
ErrorMessageService = __decorate([ __param(0, Inject(CUSTOM_ERROR_MESSAGES)),
    __metadata("design:paramtypes", [Array])
], ErrorMessageService);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ErrorMessageService, [{
        type: Injectable,
        args: [{
                providedIn: "root"
            }]
    }], function () { return [{ type: Array, decorators: [{
                type: Inject,
                args: [CUSTOM_ERROR_MESSAGES]
            }] }]; }, null); })();
export { ErrorMessageService };
