// angular-svg-round-progressbar/esm2015/round-progress/round-progress.module.js
import { NgModule } from '@angular/core';
import { RoundProgressComponent } from './round-progress.component';
import { ROUND_PROGRESS_DEFAULTS_PROVIDER } from './round-progress.config';
import * as ɵngcc0 from '@angular/core';
export class RoundProgressModule {
}
RoundProgressModule.ɵfac = function RoundProgressModule_Factory(t) { return new (t || RoundProgressModule)(); };
RoundProgressModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: RoundProgressModule });
RoundProgressModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [ROUND_PROGRESS_DEFAULTS_PROVIDER] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RoundProgressModule, [{
        type: NgModule,
        args: [{
                declarations: [RoundProgressComponent],
                exports: [RoundProgressComponent],
                providers: [ROUND_PROGRESS_DEFAULTS_PROVIDER]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RoundProgressModule, { declarations: function () { return [RoundProgressComponent]; }, exports: function () { return [RoundProgressComponent]; } }); })();
