// ngx-progressbar/bundles/ngx-progressbar-http.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-progressbar'),exports, require('@angular/core'), require('@angular/common/http'), require('rxjs/operators'), require('ngx-progressbar')) :
    typeof define === 'function' && define.amd ? define('ngx-progressbar/http', ['@angular/core','ngx-progressbar','exports', '@angular/core', '@angular/common/http', 'rxjs/operators', 'ngx-progressbar'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxProgressbar,(global['ngx-progressbar'] = global['ngx-progressbar'] || {}, global['ngx-progressbar'].http = {}), global.ng.core, global.ng.common.http, global.rxjs.operators, global['ngx-progressbar']));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, http, operators, ngxProgressbar) { 'use strict';

    var NG_PROGRESS_HTTP_CONFIG = new core.InjectionToken('ngProgressHttpConfig');

    var NgProgressInterceptor = /** @class */ (function () {
        function NgProgressInterceptor(ngProgress, config) {
            this.ngProgress = ngProgress;
            this._inProgressCount = 0;
            this._config = {
                id: 'root',
                silentApis: []
            };
            this._config = config ? Object.assign(Object.assign({}, this._config), config) : this._config;
        }
        NgProgressInterceptor.prototype.intercept = function (req, next) {
            var _this = this;
            // Ignore by request headers
            if (req.headers.has('ignoreProgressBar')) {
                return next.handle(req.clone({ headers: req.headers.delete('ignoreProgressBar') }));
            }
            // Ignore silent api requests
            if (this.checkUrl(req)) {
                return next.handle(req);
            }
            this._inProgressCount++;
            this._progressRef = this.ngProgress.ref(this._config.id);
            if (!this._progressRef.isStarted) {
                this._progressRef.start();
            }
            return next.handle(req).pipe(operators.finalize(function () {
                _this._inProgressCount--;
                if (_this._inProgressCount === 0) {
                    _this._progressRef.complete();
                }
            }));
        };
        /**
         * Check if request is silent.
         * @param req request
         */
        NgProgressInterceptor.prototype.checkUrl = function (req) {
            var url = req.url.toLowerCase();
            var found = this._config.silentApis.find(function (u) { return url.startsWith(u); });
            return !!found;
        };
NgProgressInterceptor.ɵfac = function NgProgressInterceptor_Factory(t) { return new (t || NgProgressInterceptor)(ɵngcc0.ɵɵinject(ɵngcc1.NgProgress), ɵngcc0.ɵɵinject(NG_PROGRESS_HTTP_CONFIG, 8)); };
NgProgressInterceptor.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: NgProgressInterceptor, factory: function (t) { return NgProgressInterceptor.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgProgressInterceptor, [{
        type: core.Injectable
    }], function () { return [{ type: ɵngcc1.NgProgress }, { type: undefined, decorators: [{
                type: core.Optional
            }, {
                type: core.Inject,
                args: [NG_PROGRESS_HTTP_CONFIG]
            }] }]; }, null); })();
        return NgProgressInterceptor;
    }());
    NgProgressInterceptor.ctorParameters = function () { return [
        { type: ngxProgressbar.NgProgress },
        { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [NG_PROGRESS_HTTP_CONFIG,] }] }
    ]; };

    var NgProgressHttpModule = /** @class */ (function () {
        function NgProgressHttpModule() {
        }
        NgProgressHttpModule.withConfig = function (config) {
            return {
                ngModule: NgProgressHttpModule,
                providers: [
                    { provide: NG_PROGRESS_HTTP_CONFIG, useValue: config }
                ]
            };
        };
NgProgressHttpModule.ɵfac = function NgProgressHttpModule_Factory(t) { return new (t || NgProgressHttpModule)(); };
NgProgressHttpModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgProgressHttpModule });
NgProgressHttpModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        { provide: http.HTTP_INTERCEPTORS, useClass: NgProgressInterceptor, multi: true }
    ] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgProgressHttpModule, [{
        type: core.NgModule,
        args: [{
                providers: [
                    { provide: http.HTTP_INTERCEPTORS, useClass: NgProgressInterceptor, multi: true }
                ]
            }]
    }], function () { return []; }, null); })();
        return NgProgressHttpModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NG_PROGRESS_HTTP_CONFIG = NG_PROGRESS_HTTP_CONFIG;
    exports.NgProgressHttpModule = NgProgressHttpModule;
    exports.ɵa = NgProgressInterceptor;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
