// ng-bootstrap-form-validation/esm5/lib/ng-bootstrap-form-validation.module.js
import { __decorate } from "tslib";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { FormValidationDirective } from "./Directives/form-validation.directive";
import { MessagesComponent } from "./Components/messages/messages.component";
import { ErrorMessageService } from "./Services/error-message.service";
import { CUSTOM_ERROR_MESSAGES, BOOTSTRAP_VERSION } from "./Tokens/tokens";
import { BootstrapVersion } from "./Enums/BootstrapVersion";
import { FormGroupComponent } from "./Components/form-group/form-group.component";
import { FormControlDirective } from "./Directives/form-control.directive";
import * as ɵngcc0 from '@angular/core';
var NgBootstrapFormValidationModule = /** @class */ (function () {
    function NgBootstrapFormValidationModule() {
    }
    NgBootstrapFormValidationModule_1 = NgBootstrapFormValidationModule;
    NgBootstrapFormValidationModule.forRoot = function (userOptions) {
        if (userOptions === void 0) { userOptions = {
            bootstrapVersion: BootstrapVersion.Four
        }; }
        return {
            ngModule: NgBootstrapFormValidationModule_1,
            providers: [
                {
                    provide: CUSTOM_ERROR_MESSAGES,
                    useValue: userOptions.customErrorMessages || [],
                    multi: true
                },
                {
                    provide: BOOTSTRAP_VERSION,
                    useValue: userOptions.bootstrapVersion
                },
                ErrorMessageService
            ]
        };
    };
    var NgBootstrapFormValidationModule_1;
NgBootstrapFormValidationModule.ɵfac = function NgBootstrapFormValidationModule_Factory(t) { return new (t || NgBootstrapFormValidationModule)(); };
NgBootstrapFormValidationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgBootstrapFormValidationModule });
NgBootstrapFormValidationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgBootstrapFormValidationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    FormValidationDirective,
                    FormGroupComponent,
                    MessagesComponent,
                    FormControlDirective
                ],
                exports: [
                    FormValidationDirective,
                    FormGroupComponent,
                    MessagesComponent,
                    FormControlDirective
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgBootstrapFormValidationModule, { declarations: function () { return [FormValidationDirective,
        FormGroupComponent,
        MessagesComponent,
        FormControlDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [FormValidationDirective,
        FormGroupComponent,
        MessagesComponent,
        FormControlDirective]; } }); })();
    return NgBootstrapFormValidationModule;
}());
export { NgBootstrapFormValidationModule };
