// ngx-dropzone-wrapper/esm2015/lib/dropzone.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DropzoneComponent } from './dropzone.component';
import { DropzoneDirective } from './dropzone.directive';
import * as ɵngcc0 from '@angular/core';
export class DropzoneModule {
}
DropzoneModule.ɵfac = function DropzoneModule_Factory(t) { return new (t || DropzoneModule)(); };
DropzoneModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DropzoneModule });
DropzoneModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule], CommonModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DropzoneModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [DropzoneComponent, DropzoneDirective],
                exports: [CommonModule, DropzoneComponent, DropzoneDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DropzoneModule, { declarations: function () { return [DropzoneComponent, DropzoneDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [CommonModule, DropzoneComponent, DropzoneDirective]; } }); })();
