// ngx-progressbar/esm2015/router/ng-progress-router.module.js
import { NgModule } from '@angular/core';
import { NG_PROGRESS_ROUTER_CONFIG } from './ng-progress-router.interface';
import { NgProgressRouter } from './ng-progress-router.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './ng-progress-router.service';
export class NgProgressRouterModule {
    // Inject the service to activate it
    constructor(ngProgressRouter) {
    }
    static withConfig(config) {
        return {
            ngModule: NgProgressRouterModule,
            providers: [
                { provide: NG_PROGRESS_ROUTER_CONFIG, useValue: config }
            ]
        };
    }
}
NgProgressRouterModule.ɵfac = function NgProgressRouterModule_Factory(t) { return new (t || NgProgressRouterModule)(ɵngcc0.ɵɵinject(ɵngcc1.NgProgressRouter)); };
NgProgressRouterModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgProgressRouterModule });
NgProgressRouterModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
NgProgressRouterModule.ctorParameters = () => [
    { type: NgProgressRouter }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgProgressRouterModule, [{
        type: NgModule,
        args: [{}]
    }], function () { return [{ type: ɵngcc1.NgProgressRouter }]; }, null); })();
