// ngx-loading/esm2015/lib/ngx-loading.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Inject, Injectable, Optional } from '@angular/core';
import { NgxLoadingConfig } from './ngx-loading-config';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export class NgxLoadingService {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.loadingConfig = this.config || new NgxLoadingConfig();
    }
}
NgxLoadingService.ɵfac = function NgxLoadingService_Factory(t) { return new (t || NgxLoadingService)(ɵngcc0.ɵɵinject('loadingConfig', 8)); };
NgxLoadingService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: NgxLoadingService, factory: NgxLoadingService.ɵfac, providedIn: 'root' });
/** @nocollapse */
NgxLoadingService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: ['loadingConfig',] }] }
];
/** @nocollapse */ NgxLoadingService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function NgxLoadingService_Factory() { return new NgxLoadingService(i0.ɵɵinject("loadingConfig", 8)); }, token: NgxLoadingService, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxLoadingService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: ['loadingConfig']
            }] }]; }, null); })();
if (false) {
    /** @type {?} */
    NgxLoadingService.prototype.loadingConfig;
    /** @type {?} */
    NgxLoadingService.prototype.config;
}
