// ngx-loading/esm2015/lib/ngx-loading.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgxLoadingComponent } from './ngx-loading.component';
import * as ɵngcc0 from '@angular/core';
export class NgxLoadingModule {
    /**
     * @param {?} loadingConfig
     * @return {?}
     */
    static forRoot(loadingConfig) {
        return {
            ngModule: NgxLoadingModule,
            providers: [{ provide: 'loadingConfig', useValue: loadingConfig }]
        };
    }
}
NgxLoadingModule.ɵfac = function NgxLoadingModule_Factory(t) { return new (t || NgxLoadingModule)(); };
NgxLoadingModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxLoadingModule });
NgxLoadingModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxLoadingModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NgxLoadingComponent],
                exports: [NgxLoadingComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxLoadingModule, { declarations: function () { return [NgxLoadingComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxLoadingComponent]; } }); })();
