// ngx-textarea-autosize/esm5/textarea-autosize.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AutosizeDirective } from './autosize.directive';
import * as ɵngcc0 from '@angular/core';
var TextareaAutosizeModule = /** @class */ (function () {
    function TextareaAutosizeModule() {
    }
TextareaAutosizeModule.ɵfac = function TextareaAutosizeModule_Factory(t) { return new (t || TextareaAutosizeModule)(); };
TextareaAutosizeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TextareaAutosizeModule });
TextareaAutosizeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TextareaAutosizeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AutosizeDirective],
                exports: [AutosizeDirective]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TextareaAutosizeModule, { declarations: function () { return [AutosizeDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [AutosizeDirective]; } }); })();
    return TextareaAutosizeModule;
}());
export { TextareaAutosizeModule };
